adbc.h

```diff
@@ -669,6 +675,16 @@ struct ADBC_EXPORT AdbcDriver {
                                               size_t, struct AdbcError*);
 };
 
+/// \brief An instance of an initialized database driver (API 1.1.0).
+///
+/// This provides a common interface for vendor-specific driver
+/// initialization routines. Drivers should populate this struct, and
+/// applications can call ADBC functions through this struct, without
+/// worrying about multiple definitions of the same symbol.
+struct ADBC_EXPORT AdbcDriver110 {
+  struct AdbcDriver base;
+};
```

Comment:
Why do we need a new struct? Isn't the purpose of the version to be so that we don't need a new struct?

Reply:
You can't add items to a struct without breaking binary compatibility.

Reply:
I think i'd rather break binary compatibility once here with a better solution (i guess ADBC 2.0?) than start down this path of continuously having to embed and embed and embed until we end up with stuff like `driver->base.base.base.base.....`

Reply:
If there's appetite for 2.0.0, we could do that. Most of the proposals so far are not drastic changes: https://github.com/apache/arrow-adbc/milestone/3 except for possibly arbitrarily nested database schemas (which doesn't fit well into the current nested data hierarchy from GetObjects, since Arrow doesn't have recursive types).

We could use the driver manager to still provide compatibility if we want to do 2.0.0.

Reply:
I'd prefer reserving something for future expansion rather than the embedding.

Reply:
@paleolimbot wonder if you have any thoughts here?

Reply:
We could also do something where based on the passed in version we could assume the presence/absence of new pointers and avoid accessing the "new part" of the struct, UCX-style (as mentioned below). This is rather janky, but would be OK if we assume this mechanism is primarily for the driver/driver manager to coordinate.

Reply:
How does the approach here look? https://github.com/apache/arrow-adbc/pull/692

If it works, I'll figure out how to set up all the compatibility tests.

Reply:
I actually like that approach better. Will add comments to that PR